src/display/annotation_layer.js

```diff
@@ -374,19 +386,82 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
      * @returns {HTMLSectionElement}
      */
     render: function WidgetAnnotationElement_render() {
-      var content = document.createElement('div');
-      content.textContent = this.data.fieldValue;
+      var isReadonly = this.isFlagSet(READONLY_BIT);
+      this.container.disabled = isReadonly;
+      return this.container;
+    },
+
+    /**
+     *
+     * @protected
+     * @memberof WidgetAnnotationElement
+     * @returns {boolean}
+     */
+    isFlagSet: function WidgetAnnotationElement_isFlagSet(pos) {
+      // Note: pos is 1 based index
+      return !!(this.data.fieldFlags & (1 << (pos - 1)));
+    }
+  });
+
+  return WidgetAnnotationElement;
+})();
+
+/**
+ * @class
+ * @alias TextWidgetAnnotationElement
+ */
+var TextWidgetAnnotationElement =
+    (function TextWidgetAnnotationElementClosure() {
+  function TextWidgetAnnotationElement(parameters) {
+    WidgetAnnotationElement.call(this, parameters);
+  }
+
+  Util.inherit(TextWidgetAnnotationElement, WidgetAnnotationElement, {
+    /**
+     *
+     * @public
+     * @memberof TextWidgetAnnotationElement
+     * @returns {HTMLSectionElement}
+     */
+    render: function TextWidgetAnnotationElement_render() {
+      var container = WidgetAnnotationElement.prototype.render.call(this);
+
+      var isMultiline = this.isFlagSet(MULTILINE_BIT);
+
+      var content;
+
+      if (isMultiline) {
+        content = document.createElement('textarea');
+      } else {
+        content = document.createElement('input');
+        content.type = 'text';
+      }
+
+      content.value = this.data.fieldValue;
       var textAlignment = this.data.textAlignment;
       content.style.textAlign = ['left', 'center', 'right'][textAlignment];
-      content.style.verticalAlign = 'middle';
-      content.style.display = 'table-cell';
+      if (this.data.maxLen !== null) {
+        content.maxLength = this.data.maxLen;
+      }
 
       var font = (this.data.fontRefName ?
         this.page.commonObjs.getData(this.data.fontRefName) : null);
       this._setTextStyle(content, font);
 
-      this.container.appendChild(content);
-      return this.container;
+      if (!isMultiline && !('fontSize' in this.data)) {
+        // hack to guess font size base on content hight
```

Comment:
An example file that needs this hack is https://drive.google.com/file/d/0B6YhVF9p-y2AYzRNLTE5Z2dFWHM/view?usp=sharing
Without it some of the fields will be cut because the fonts are too large